subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ForcedDependencyMetadataWrapper.java

```diff
@@ -75,12 +76,20 @@ public ModuleDependencyMetadata withEndorseStrictVersions(boolean endorse) {
 
     @Override
     public DependencyMetadata withTarget(ComponentSelector target) {
-        return new ForcedDependencyMetadataWrapper((ModuleDependencyMetadata) delegate.withTarget(target));
+        DependencyMetadata dependencyMetadata = delegate.withTarget(target);
+        if (dependencyMetadata instanceof DefaultProjectDependencyMetadata) {
+            return ((DefaultProjectDependencyMetadata) dependencyMetadata).forced();
+        }
+        return new ForcedDependencyMetadataWrapper((ModuleDependencyMetadata) dependencyMetadata);
     }
 
     @Override
     public DependencyMetadata withTargetAndArtifacts(ComponentSelector target, List<IvyArtifactName> artifacts) {
-        return new ForcedDependencyMetadataWrapper((ModuleDependencyMetadata) delegate.withTargetAndArtifacts(target, artifacts));
+        DependencyMetadata dependencyMetadata = delegate.withTargetAndArtifacts(target, artifacts);
+        if (dependencyMetadata instanceof DefaultProjectDependencyMetadata) {
```

Comment:
The structure of the various implementations of `DependencyMetadata` grew into quite a mess over time. It's a mix of inheritance, multiple interfaces and delegation. There is no consistent pattern in that right now. Hence I went for the "simple" fix here that addresses the one special case that was broken, instead of introducing even more abstraction/delegation.